### PR TITLE
fix(optimizer): Force early alias expansion in BQ & CH

### LINKED ISSUE #3687: wrong qualify_columns on generated columns
**Before you file an issue**
- Make sure you specify the "read" dialect eg. `parse_one(sql, read="spark")`: Done
- Make sure you specify the "write" dialect eg. `ast.sql(dialect="duckdb")`: Done
- Check if the issue still exists on main: still exist

**Fully reproducible code snippet**

Prepare data:
```python
from sqlglot import parse_one, exp
from sqlglot.optimizer.qualify_columns import qualify_columns

sql ="""with data as (
  select 1 as id, 2 as my_id, 'a' as name, 'de' as full_name
)
select
  id as my_id,
  concat(id, name) as full_name
from data
group by my_id, full_name"""

schema = {
"catalog": {
      "schema1": {
          "customer": {
              "id": "INTEGER",
          },
      }
  }
}
```

Transpiler from bigquery to trino:
```python
smt = parse_one(sql, read="bigquery")
print(qualify_columns(smt, schema=schema).sql(dialect="trino", pretty=True))
```
Output:
```sql
WITH data AS (
  SELECT
    1 AS id,
    2 AS my_id,
    'a' AS name,
    'de' AS full_name
)
SELECT
  data.id AS my_id,
  CONCAT(CAST(data.id AS VARCHAR), CAST(data.name AS VARCHAR)) AS full_name
FROM data
GROUP BY
  data.my_id,
  data.full_name
```

The problem of the output is it can not be run on trino, because `data.id` and `data.name` is not an aggregate expression or appear in GROUP BY clause.

Additional test:
The output will expand correctly if we no not provide schema:
```python
smt = parse_one(sql, read="bigquery")
print(qualify_columns(smt, schema=None).sql(dialect="trino", pretty=True))

"""
WITH data AS (
  SELECT
    1 AS id,
    2 AS my_id,
    'a' AS name,
    'de' AS full_name
)
SELECT
  data.id AS my_id,
  CONCAT(CAST(data.id AS VARCHAR), CAST(data.name AS VARCHAR)) AS full_name
FROM data
GROUP BY
  data.id,
  CONCAT(CAST(data.id AS VARCHAR), CAST(data.name AS VARCHAR))
"""
```

The output will break again if I disable expand_alias_ref: 
```python
smt = parse_one(sql, read="bigquery")
print(qualify_columns(smt, schema=None, expand_alias_refs=False).sql(dialect="trino", pretty=True))
```

### Solution
I have made a patch to fix my case, but I think we can have better solution to fix this.

https://github.com/viplazylmht/sqlglot/blob/644bf67720442f964ffc4606693dcfc1b87ec6e6/sqlglot/optimizer/qualify_columns.py#L423-L431

**Official Documentation**
Please include links to official SQL documentation related to your issue.

### PR DESCRIPTION
**Pull request by VaggelisD**
Fixes #3687

Take the following query:

```SQL
WITH data AS (
SELECT
    1 AS id,
    2 AS my_id
)
SELECT
    id AS my_id
FROM
    data
WHERE
    my_id = 1
GROUP BY
    my_id,
HAVING
    my_id = 1
```

In most dialects this is an incorrect query as `GROUP BY my_id` is resolved to `GROUP BY data.my_id` which is not aggregated (the overlapping alias in the projection list is resolved later in the order of operations). 

However, the exceptions to this rule which _can_ process this query are the following:
- BigQuery, which only forwards the alias only to `GROUP BY` & `HAVING` i.e resolves to:
```SQL
WITH data AS (...)
WHERE data.my_id = 1 GROUP BY data.id HAVING data.id = 1
```

- Clickhouse, which forwards the alias across the query i.e resolves to:
```SQL
WITH data AS (...)
WHERE data.id = 1 GROUP BY data.id HAVING data.id = 1
```

This PR aims to fix this by forcing the alias expansion before `_qualify_columns()` for these cases, as the 2nd call to `_expand_alias_refs()` does not expand aliases on qualified columns.

### CODE REVIEW

File: tests/test_optimizer.py [hunk 1]
```diff
@@ -473,15 +473,35 @@ def test_expand_alias_refs(self):
...
                 infer_schema=False,
             ).sql(),
             "SELECT CAST(x AS INT) AS y FROM z AS z",
         )
 
+        # BigQuery expands overlapping alias only for GROUP BY + HAVING
+        sql = "WITH data AS (SELECT 1 AS id, 2 AS my_id, 'a' AS name, 'b' AS full_name) SELECT id AS my_id, CONCAT(id, name) AS full_name FROM data WHERE my_id = 1 GROUP BY my_id, full_name HAVING my_id = 1"
+        self.assertEqual(
+            optimizer.qualify_columns.qualify_columns(
+                parse_one(sql),
```
Comment (georgesittas) [inline tests/test_optimizer.py:490]:
```suggestion
                parse_one(sql, dialect="bigquery"),
```

File: tests/test_optimizer.py [hunk 2]
```diff
@@ -473,15 +473,35 @@ def test_expand_alias_refs(self):
...
+                schema=MappingSchema(schema=unused_schema, dialect="bigquery"),
+            ).sql(),
+            "WITH data AS (SELECT 1 AS id, 2 AS my_id, 'a' AS name, 'b' AS full_name) SELECT data.id AS my_id, CONCAT(data.id, data.name) AS full_name FROM data WHERE data.my_id = 1 GROUP BY data.id, CONCAT(data.id, data.name) HAVING data.id = 1",
+        )
+
+        # Clickhouse expands overlapping alias across the entire query
+        self.assertEqual(
+            optimizer.qualify_columns.qualify_columns(
+                parse_one(sql),
```
Comment (georgesittas) [inline tests/test_optimizer.py:499]:
```suggestion
                parse_one(sql, dialect="clickhouse"),
```

**Review by georgesittas — APPROVED** [review]

File: sqlglot/optimizer/qualify_columns.py [hunk 3]
```diff
@@ -60,8 +60,8 @@ def qualify_columns(
         _pop_table_column_aliases(scope.derived_tables)
         using_column_tables = _expand_using(scope, resolver)
 
-        if schema.empty and expand_alias_refs:
-            _expand_alias_refs(scope, resolver)
+        if (schema.empty or dialect.FORCE_EARLY_ALIAS_REF_EXPANSION) and expand_alias_refs:
+            _expand_alias_refs(scope, resolver, expand_only_groupby=(dialect == "bigquery"))
```
Comment (viplazylmht) [inline sqlglot/optimizer/qualify_columns.py:64]: Hello @VaggelisD 

Should we provide the `expand_only_groupby` via qualify_columns parameters? (we can infer it with your logic if user provide None).  

The context is if I extend the Bigquery dialect to MyBigquery dialect, this change will not work anymore, so I want to force it using a parameter passed to the function caller instead.

Comment (georgesittas) [reply sqlglot/optimizer/qualify_columns.py:64]: Can't you just set the flag in your dialect?

Comment (viplazylmht) [reply sqlglot/optimizer/qualify_columns.py:64]: Sure, `expand_only_groupby` inside the dialect is a good idea.

Comment (georgesittas) [reply sqlglot/optimizer/qualify_columns.py:64]: Ah you were talking about the expand only group logic, I completely missed it sorry. We could probably parameterize it, cc @VaggelisD

Comment (georgesittas) [reply sqlglot/optimizer/qualify_columns.py:64]: https://github.com/tobymao/sqlglot/pull/3702

**Pull request merged**